src/core/mount.c

```diff
@@ -759,7 +759,7 @@ static int mount_coldplug(Unit *u) {
         assert(m);
         assert(m->state == MOUNT_DEAD);
 
-        if (m->deserialized_state != m->state)
+        if (m->deserialized_state != m->state && m->deserialized_state != MOUNT_MOUNTING)
                 new_state = m->deserialized_state;
         else if (m->from_proc_self_mountinfo)
                 new_state = MOUNT_MOUNTED;
```

Comment:
Maybe this:
```c
if (m->deserialized_state != m->state)
        new_state = m->deserialized_state;
if (m->from_proc_self_mountinfo && new_state == MOUNT_MOUNTING)
        new_state = MOUNT_MOUNTED;
```
This means we were mounting something before reloading, now it is mounted, the new_state should keep consistent with proc_self_mountinfo.
@yuwata @poettering PTAL.

Reply:
@Geass-LL That should be better. But how about the other states? Especially when `new_state` is `MOUNT_DEAD`, `MOUNT_MOUNTING_DONE`, or `MOUNT_REMOUNTING`.
Also, if `m->from_proc_self_mountinfo` is `NULL`, then the state should be also adjusted, I guess.

Reply:
@yuwata I thought we can copy the logic in `mount_process_proc_self_mountinfo` yesterday. 
```c
MOUNT_DEAD/MOUNT_FAILED -> MOUNT_MOUNTED
MOUNT_MOUNTING_DONE -> MOUNT_MOUNTING_DONE
MOUNT_MOUNTING/MOUNT_REMOUNTING -> MOUNT_MOUNTING_DONE
```
Because at that time, I didn't konw systemd would process `/proc/self/mountinfo`  just after receiving sigchld of mount.
<https://github.com/systemd/systemd/blob/f0390fa03456ecb4d5f1482bd20a28ea98ae4d58/src/core/mount.c#L1376-L1390>
Now I'm adding some log to `mount_process_proc_self_mountinfo` to see why systemd didn't change the mount->state from MOUNT_MOUNTING to MOUNT_MOUNTING_DONE.

Reply:
Thank you all for your discussion and I look forward to a better fix.